File: 백준/Silver/28215. 대피소/대피소.py
import sys
input = sys.stdin.readline

n, K = map(int, input().split())

home = [list(map(int, input().split())) for _ in range(n)]

ans = sys.maxsize

for i in range(n):
    if K > 1 :
        for j in range(i+1, n):
            # print(f"i is {i}, j is {j}")
            if K > 2:
                for k in range(j+1, n):
                    max_dist = 0
                    for x in range(n):
                        min_dist = sys.maxsize
                        min_dist = min(min_dist, abs(home[i][0] - home[x][0]) + abs(home[i][1] - home[x][1]))
                        min_dist = min(min_dist, abs(home[j][0] - home[x][0]) + abs(home[j][1] - home[x][1]))
                        min_dist = min(min_dist, abs(home[k][0] - home[x][0]) + abs(home[k][1] - home[x][1]))
                        max_dist = max(max_dist, min_dist)
                    ans = min(max_dist, ans)
            else :
                max_dist = 0
                for x in range(n):
                    min_dist = sys.maxsize
                    min_dist = min(min_dist, abs(home[i][0] - home[x][0]) + abs(home[i][1] - home[x][1]))
                    min_dist = min(min_dist, abs(home[j][0] - home[x][0]) + abs(home[j][1] - home[x][1]))
                    max_dist = max(max_dist, min_dist)
                ans = min(max_dist, ans)
                    
                    # print(f"x is {x}")
                    # # print(f"max_dist: {max_dist}")
                    # print(f"min_dist: {min_dist}")
                    # print()

    else :
        max_dist = 0
        for x in range(n): 
            min_dist = sys.maxsize
            min_dist = min(min_dist, abs(home[i][0] - home[x][0]) + abs(home[i][1] - home[x][1]))
            max_dist = max(max_dist, min_dist)
        ans = min(max_dist, ans)
    

print(ans)

File: 백준/Silver/28215. 대피소/대피소.cc
#include <bits/stdc++.h>
using namespace std;

int n, k;
vector<pair<int, int>> home;

int main() {
    cin >> n >> k;
    for (int i = 0; i < n; i++) {
        int x, y;
        cin >> x >> y;
        home.push_back({x, y});
    }

    int ans = INT_MAX;
    for (int i = 0; i < n; i++) {
        if (k > 1) {
            for (int j = i+1; j < n; j++) {
                if (k > 2) {
                    for (int k = j+1; k < n; k++) {
                        int max_dist = 0;
                        for (int x = 0; x < n; x++) {
                            int min_dist = INT_MAX;
                            min_dist = min(min_dist, abs(home[i].first - home[x].first) + abs(home[i].second - home[x].second));
                            min_dist = min(min_dist, abs(home[j].first - home[x].first) + abs(home[j].second - home[x].second));
                            min_dist = min(min_dist, abs(home[k].first - home[x].first) + abs(home[k].second - home[x].second));
                            max_dist = max(max_dist, min_dist);
                        }
                    }
                }
                else {
                    int max_dist = 0;
                    for (int x = 0; x < n; x++) {
                        int min_dist = INT_MAX;
                        min_dist = min(min_dist, abs(home[i].first - home[x].first) + abs(home[i].second - home[x].second));
                        min_dist = min(min_dist, abs(home[j].first - home[x].first) + abs(home[j].second - home[x].second));
                        max_dist = max(max_dist, min_dist);
                    }
                    ans = min(max_dist, ans);
                }
            }
        }
        else {
            int max_dist = 0;
            for (int x = 0; x < n; x++) {
                int min_dist = INT_MAX;
                min_dist = min(min_dist, abs(home[i].first - home[x].first) + abs(home[i].second - home[x].second));
                max_dist = max(max_dist, min_dist);
            }
            ans = min(max_dist, ans);
        }
    }

    cout << ans;


    return 0;
}